flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java

```diff
@@ -264,25 +270,43 @@ public boolean executeInitialization(String content) {
      * terminal.
      */
     private void executeInteractive() {
-        isRunning = true;
-        LineReader lineReader = createLineReader(terminal);
-
         // make space from previous output and test the writer
         terminal.writer().println();
         terminal.writer().flush();
 
         // print welcome
         terminal.writer().append(CliStrings.MESSAGE_WELCOME);
 
+        LineReader lineReader = createLineReader(terminal);
+        getAndExecuteStatements(lineReader, ExecutionMode.INTERACTIVE_EXECUTION);
+    }
+
+    private boolean getAndExecuteStatements(LineReader lineReader, ExecutionMode mode) {
         // begin reading loop
+        boolean exitOnFailure = !mode.equals(ExecutionMode.INTERACTIVE_EXECUTION);
+        isRunning = true;
         while (isRunning) {
             // make some space to previous command
             terminal.writer().append("\n");
             terminal.flush();
 
-            String line;
+            Optional<Operation> parsedOperation = Optional.empty();
             try {
-                line = lineReader.readLine(prompt, null, inputTransformer, null);
+                String line = lineReader.readLine(prompt, null, inputTransformer, null);
+                if (line.trim().isEmpty()) {
+                    continue;
+                }
+                parsedOperation = parser.getParsedOperation();
```

Comment:
I had a bit of hard time to understand how state propagation works here. I think we can do better and make it more clear: You can make`SqlMultiLineParser` stateless, and propagate every parsing result wrapping `parseException` and `parsedOperation` within `SqlArgumentList` (perhaps also rename this class), and then here you can access the last parsed `SqlArgumentList` using `lineReader.getParsedLine()`.

With this you don't need to check any correctness of state propagation (with the below checkArgument) and you don't need to keep the latest command line. And it makes the code overrall simpler to understand.

Reply:
I see, then let's keep it as it is. Perhaps just comment the method `getParsedOperation` and a bit the code here to explain the relationship and the state sharing between these two objects